File: Week 3/type_checking.cpp
//Static type checking in C++

#include <iostream>
#include <string>
using namespace std;

int main(){
    int testVar = "Hello";       // ❌ Type error: can't assign a string to an int
    int result = 15 * "abc";     // ❌ Type error: can't multiply int by string
    string testResult = 14 * 150;  // ❌ Type error: can't assign int to string
    return 0;
}

File: Week 3/type_checking.py
#Dynamic type checking in Python 

test_var = "Hello"        # allowed
result = 15 * "abc"       # allowed → result is "abc" repeated 15 times
test_result = 14 * 150    # allowed → result is 2100

# But if we try incompatible operations:
x = "Hello" + 5           # ❌ TypeError at runtime

#Difference between implicit variable declaration are dynamic typing
#Implicit variable declaration:
x = 10      # x is created implicitly as an int
y = "hello" # y is created implicitly as a string

#Dynamic typing
x = 10       # x is an int
x = "hello"  # now x is a string — allowed in dynamically typed languages